File: codes/golang/743.网络延迟时间.go

import (
	"container/heap"
	"math"
)

func networkDelayTime(times [][]int, n int, k int) int {
	// 节点编号是从 1 开始的，所以要一个大小为 n + 1 的邻接表
	graph := make([][][]int, n+1)
	for i := 1; i <= n; i++ {
		graph[i] = make([][]int, 0)
	}

	// 构造图
	for _, edge := range times {
		from := edge[0]
		to := edge[1]
		weight := edge[2]
		// from -> [ [to, weight], ... ]
		// 邻接表存储图结构，同时存储权重信息
		graph[from] = append(graph[from], []int{to, weight})
	}

	// 启动 dijkstra 算法计算以节点 k 为起点到其他节点的最短路径
	distTo := dijkstra(k, graph)

	// 找到最长的那一条最短路径
	res := 0
	for i := 1; i < len(distTo); i++ {
		if distTo[i] == math.MaxInt32 {
			// 有节点不可达，返回 -1
			return -1
		}
		res = max(res, distTo[i])
	}
	return res
}

type State struct {
	// 图节点的 id
	id int
	// 从 start 节点到当前节点的距离
	distFromStart int
}

type PriorityQueue []*State

func (pq PriorityQueue) Len() int { return len(pq) }
func (pq PriorityQueue) Less(i, j int) bool {
	return pq[i].distFromStart < pq[j].distFromStart
}
func (pq PriorityQueue) Swap(i, j int) { pq[i], pq[j] = pq[j], pq[i] }

func (pq *PriorityQueue) Push(x interface{}) {
	item := x.(*State)
	*pq = append(*pq, item)
}

func (pq *PriorityQueue) Pop() interface{} {
	old := *pq
	n := len(old)
	item := old[n-1]
	old[n-1] = nil // avoid memory leak
	*pq = old[0 : n-1]
	return item
}

// 输入一个起点 start，计算从 start 到其他节点的最短距离
func dijkstra(start int, graph [][][]int) []int {
	// 定义：distTo[i] 的值就是起点 start 到达节点 i 的最短路径权重
	distTo := make([]int, len(graph))
	for i := 1; i < len(graph); i++ {
		distTo[i] = math.MaxInt32
	}
	// base case，start 到 start 的最短距离就是 0
	distTo[start] = 0

	// 优先级队列，distFromStart 较小的排在前面
	pq := make(PriorityQueue, 0)
	heap.Init(&pq)
	// 从起点 start 开始进行 BFS
	heap.Push(&pq, &State{id: start, distFromStart: 0})

	for pq.Len() > 0 {
		curState := heap.Pop(&pq).(*State)
		curNodeID := curState.id
		curDistFromStart := curState.distFromStart

		if curDistFromStart > distTo[curNodeID] {
			continue
		}

		// 将 curNode 的相邻节点装入队列
		for _, neighbor := range graph[curNodeID] {
			nextNodeID := neighbor[0]
			distToNextNode := distTo[curNodeID] + neighbor[1]
			// 更新 dp table
			if distTo[nextNodeID] > distToNextNode {
				distTo[nextNodeID] = distToNextNode
				heap.Push(&pq, &State{id: nextNodeID, distFromStart: distToNextNode})
			}
		}
	}
	return distTo
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: codes/python/743.网络延迟时间.py
import queue
from typing import List


class State():
    def __init__(self, node_id: int, dist_from_start: int):
        self.node_id = node_id
        self.dist_from_start = dist_from_start

    def __lt__(self, other):
        return self.dist_from_start < other.dist_from_start

def custom_compare(item: State) -> int:
    return item.dist_from_start

def dijkstra(graph: List[List[tuple]], start: int) -> List[int]:
    # 定义一个列表，用于存在start节点到每个节点的最短路径
    dist_to_list = [float("inf") for _ in range(len(graph))]
    dist_to_list[start] = 0

    # 定义一个优先级队列， 从拥有最短路径的节点开始遍历
    pq  = queue.PriorityQueue()
    pq.put(State(start, 0))

    # 从队列中取出最短路径的节点，然后更新其相邻节点的最短路径
    while not pq.empty():
        cur_node: State = pq.get()
        cur_node_id = cur_node.node_id
        cur_node_dist_from_start = cur_node.dist_from_start

        # 如果当前节点的路径 > 当前的最短路径，那么就忽略掉
        if cur_node_dist_from_start > dist_to_list[cur_node_id]:
            continue

        # 遍历当前节点的相邻节点，更新其最短路径
        edgs = graph[cur_node_id]
        for edg in edgs:
            next_node_id = edg[0]
            edg_weight = edg[1]
            # 如果当前节点 + 边的权重要比相邻节点的最短路径要小，那么就更新其最短路径, 并将其加入到优先级队列中
            if cur_node_dist_from_start + edg_weight < dist_to_list[next_node_id]:
                dist_to_list[next_node_id] = cur_node_dist_from_start + edg_weight
                pq.put(State(next_node_id, dist_to_list[next_node_id]))

    # 最后返回每个节点的最短路径
    return dist_to_list


class Solution:
    def networkDelayTime(self, times: List[List[int]], n: int, k: int) -> int:
        # 第一步: 首先构造图
        graph = []
        for _ in range(n+1):
            graph.append([])
        for time in times:
            u = time[0]
            v = time[1]
            w = time[2]
            graph[u].append((v, w))
        
        # 第二步: 基于上面的图开始使用dijkstra构造每个节点的最短路径(从start节点到当前节点)
        dist_to_list = dijkstra(graph, k)
        # 第三步: 取其中一个路径最大的节点就是访问完所有节点的最大网络延迟时间
        max_time = 0
        for i in range(1, len(dist_to_list)):
            if dist_to_list[i] == float("inf"):
                return -1
            max_time = max(max_time, dist_to_list[i])
        return max_time            